adding attribute source to EPC attributes

YaphetKG opened the pull request:
This PR sets `attribute_source` for provenance attributes to the current plater instance name.  (https://github.com/TranslatorSRI/Plater/issues/66) 

Eg of an output 
```
    "knowledge_graph": {
      "nodes": {
        "NCBIGene:1714": {
          "categories": ...
          "name": "DiGeorge sequence",
          "attributes": [
            {
              "attribute_type_id": "biolink:provided_by",
              "value": [
                "merge_nodes.jsonl"
              ],
              "value_type_id": "EDAM:data_0006",
              "original_attribute_name": "provided_by",
              "value_url": null,
              "attribute_source": null,
              "description": null,
              "attributes": null
            },....
            {
              "attribute_type_id": "biolink:aggregator_knowledge_source",
              "value": "infores:automat.notspecified",
              "value_type_id": "biolink:InformationResource",
              "original_attribute_name": "biolink:aggregator_knowledge_source",
              "value_url": null,
              "attribute_source": "infores:automat.notspecified",
              "description": null,
              "attributes": null
            }
          ]
        },
        "MONDO:0009010":  ...
      },
      "edges": {
        "b9614111fb4ce63393463c0b09474341": {
          "subject": "NCBIGene:1714",
          "object": "MONDO:0009010",
          "predicate": "biolink:has_phenotype",
          "attributes": [
            {
              "attribute_type_id": "biolink:knowledge_source",
              "value": [
                "merge_edges.jsonl"
              ],
              "value_type_id": "EDAM:data_0006",
              "original_attribute_name": "knowledge_source",
              "value_url": null,
              "attribute_source": null,
              "description": null,
              "attributes": null
            },
            {
              "attribute_type_id": "biolink:aggregator_knowledge_source",
              "value": [
                "infores:sri-reference-kg"
              ],
              "value_type_id": "biolink:InformationResource",
              "original_attribute_name": "biolink:aggregator_knowledge_source",
              "value_url": null,
              "attribute_source": "infores:automat.notspecified",
              "description": null,
              "attributes": null
            },
            {
              "attribute_type_id": "biolink:relation",
              "value": "RO:0002200",
              "value_type_id": "EDAM:data_0006",
              "original_attribute_name": "relation",
              "value_url": null,
              "attribute_source": null,
              "description": null,
              "attributes": null
            },
            {
              "attribute_type_id": "biolink:aggregator_knowledge_source",
              "value": "infores:automat.notspecified",
              "value_type_id": "biolink:InformationResource",
              "original_attribute_name": "biolink:aggregator_knowledge_source",
              "value_url": null,
              "attribute_source": "infores:automat.notspecified",
              "description": null,
              "attributes": null
            }
          ]
        }
      }
    }
```
( infores:automat.notspecified  is the default local server kp name , in prod this would be an actual instance name)

cbizon commented:
Thanks @YaphetKG .   There are a few things that I'd like to fix here.  (Most of these are leftover issues, I think the changes you made address the actual issue)

1. I don't think we should have any provenance attributes on nodes.  Neither provided_by nor aggregated_knowledge_source.
2. This attribute: 
```"attribute_type_id": "biolink:knowledge_source",
              "value": [
                "merge_edges.jsonl"
              ],``` 
   is meaningless and should be removed
3. We should have an original or primary knowledge source on this edge.   
4. For one aggregator_knowledge_source value is a string and for one it's a list.  I'm guessing both are allowed?  We might as well standardize one way or the other.

YaphetKG commented:
Hi Chris, 
Here are some changes, and have a question on 3 , 

1. Provenance is now added to edges only 

2.  KGX adds in those attributes, but we maintain a skip_attr.json list that skips attributes with certain names. I have added knowledge_source and provided_by attributes to that list by default here. (side note: we already mount this file on the deployments but now it also in the repo as a default).

3. The previous example was from biolink neo4j backend ,  another example from ctd contains edge attributes such as 
```
"e6afd7049afd97cedd69ae64683ce0e2": {
          "subject": "PUBCHEM.COMPOUND:10836",
          "object": "NCBIGene:834",
          "predicate": "biolink:increases_degradation_of",
          "attributes": [
            { # coming from the graph
              "attribute_type_id": "biolink:original_knowledge_source",
              "value": [
                "infores:ctd"
              ],
              "value_type_id": "biolink:InformationResource",
              "original_attribute_name": "biolink:original_knowledge_source",
              "value_url": null,
              "attribute_source": "infores:automat.notspecified",
              "description": null,
              "attributes": null
            },
            {
              "attribute_type_id": "biolink:relation",
              "value": "CTD:increases_degradation_of",
              "value_type_id": "EDAM:data_0006",
              "original_attribute_name": "relation",
              "value_url": null,
              "attribute_source": null,
              "description": null,
              "attributes": null
            },
            {   # added on the fly depending on the plater tag (in prod infores:automat.notspecified would be infores:automat-ctd)
              "attribute_type_id": "biolink:aggregator_knowledge_source",
              "value": [
                "infores:automat.notspecified"
              ],
              "value_type_id": "biolink:InformationResource",
              "original_attribute_name": "biolink:aggregator_knowledge_source",
              "value_url": null,
              "attribute_source": "infores:automat.notspecified",
              "description": null,
              "attributes": null
            }
          ]
        },
        "f459261a9e24d1ca9420bd24f820363b": { ... }
```
The biolink example had an attirbute `biolink:aggregator_knowledge_source` twice because there is one coming from the graph and another added on the fly. The on the fly addition is done for all that are served through plater. Do we want to drop / change what we add on the fly to primary instead of aggregate ?

4. Added some code to convert all provenance values to lists.

cbizon commented:
OK, these look good.   For 3, there's no problem with having 2 aggregators, the problem is that there's no primary or original source.   However, I think that this is not Plater's problem.  As you say, that should be coming from the graph file itself, so I will put a ticket on Data Services for that.
